src/webview/html/messageHeaderAsHtml.js

```diff
@@ -57,45 +60,49 @@ export default (
     const topicHtml = renderSubject(item);
 
     return template`
-<div class="header-wrapper stream-header" data-msg-id="${item.id}">
+<div class="header-wrapper header stream-header" data-msg-id="${item.id}">
   <div class="header stream-text"
        style="color: ${textColor};
               background: ${backgroundColor}"
        data-narrow="${streamNarrowStr}">
     # ${item.display_recipient}
   </div>
-  <div class="header topic-text" data-narrow="${topicNarrowStr}">
-    $!${topicHtml}
+  <div class="header-wrapper header topic-content" data-narrow="${topicNarrowStr}">
+    <div class="topic-text">$!${topicHtml}</div>
+    <div class="human-date">$!${humanDate(new Date(item.timestamp * 1000))}</div>
   </div>
 </div>
     `;
   }
 
-  if (
-    item.type === 'private'
-    && !isPrivateNarrow(narrow)
-    && !isGroupNarrow(narrow)
-    && !isTopicNarrow(narrow)
-  ) {
-    const recipients =
-      item.display_recipient.length === 1 && item.display_recipient[0].email === ownEmail
-        ? item.display_recipient
-        : item.display_recipient.filter(r => r.email !== ownEmail);
+  if (item.type === 'private') {
+    let recipientsStr;
+    let privateNarrowStr;
+
+    if (isSpecialNarrow(narrow) || isHomeNarrow(narrow) || isGroupNarrow(narrow)) {
+      const recipients =
+        item.display_recipient.length === 1 && item.display_recipient[0].email === ownEmail
+          ? item.display_recipient
+          : item.display_recipient.filter(r => r.email !== ownEmail);
+
+      recipientsStr = recipients
+        .map(r => r.full_name)
+        .sort()
+        .join(', ');
 
-    const narrowObj =
-      recipients.length === 1
-        ? privateNarrow(recipients[0].email)
-        : groupNarrow(recipients.map(r => r.email));
-    const privateNarrowStr = JSON.stringify(narrowObj);
+      const narrowObj =
+        recipients.length === 1
+          ? privateNarrow(recipients[0].email)
+          : groupNarrow(recipients.map(r => r.email));
+      privateNarrowStr = JSON.stringify(narrowObj);
+    }
 
     return template`
-<div class="header-wrapper private-header header"
+<div class="header-wrapper header private-header"
      data-narrow="${privateNarrowStr}"
      data-msg-id="${item.id}">
-  ${recipients
-    .map(r => r.full_name)
-    .sort()
-    .join(', ')}
+     <div class="topic-text">$!${recipientsStr}</div>
+     <div class="human-date">$!${humanDate(new Date(item.timestamp * 1000))}</div>
```

Comment:
Dangerous -- this is really important!

Same thing I mentioned in a comment in the last round:
https://github.com/zulip/zulip-mobile/pull/3435#discussion_r271046563

Essentially this is a type error -- you are taking content that is plain text, and using it in a place that expects HTML. The only correct way to do that is to *encode* the text as HTML.

Otherwise when some user changes their name in Zulip to something like, say, `<script>alert("lol pwned");</script>`, and then sends you a PM -- something very much the wrong thing happens. :stuck_out_tongue_winking_eye: 

The term for this is "HTML injection". Here's one reasonable piece of background on the issue:
https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.md
In that page's terms, "Rule # 1" is the applicable rule here.

In our `template` function which we're using here, the normal way to encode text as HTML is to use a plain `${...}`. The `!` in `$!$` is a bright red warning sign! Please heed the warning sign.

Reply:
Thanks for the explanation and the link (helped a lot more than understanding HTML injection too :) )!

I was actually aware of HTML injection being a real thing, and I had fixed this issue in WD before (as explained in the comment above) must have forgotten to commit it for some reason.

(That link is very resourceful!)